src/stream/src/executor/dispatch.rs

```diff
@@ -228,85 +229,101 @@ impl DispatchExecutorInner {
         Ok(())
     }
 
-    /// For `Add` and `Update`, update the dispatchers before we dispatch the barrier.
+    /// Update the dispatcher in this executor with `update`.
+    fn update_dispatcher<const PRE: bool>(&mut self, update: &ProstDispatcherUpdate) -> Result<()> {
+        let dispatcher = self
+            .dispatchers
+            .iter_mut()
+            .find(|d| d.dispatcher_id() == update.dispatcher_id)
+            .unwrap_or_else(|| panic!("dispatcher {} not found", update.dispatcher_id));
+
+        if PRE {
+            // Update the dispatcher BEFORE we actually dispatch this barrier. We'll only add the
+            // new outputs.
+            let outputs: Vec<_> = update
+                .added_downstream_actor_id
+                .iter()
+                .map(|&id| new_output(&self.context, self.actor_id, id))
+                .try_collect()?;
+            dispatcher.add_outputs(outputs);
+        } else {
+            // Update the dispatcher AFTER we dispatch this barrier. We'll remove some outputs and
+            // finally update the hash mapping.
```

Comment:
Why doesn't it require updating the hash mapping to add outputs? 🤔

Reply:
Got it 🤣. So actually `pre` and `post` will both be called, they are for different _parts_ of an Update (instead of different _cases_ of Update)